### ISSUE TITLE
baseball.py example freezing on travis intermittently

### ISSUE THREAD

**Issue by neerajprad** [uber/pyro#1622]
I noticed in https://github.com/uber/pyro/pull/1619 that the baseball.py example in `test_examples.py` froze midway and the build terminated due to `no output received`. 

My guess is that this is an issue with python multiprocessing not playing well on travis. If we continue to see this, I will default to running a single chain on travis.

**Comment by fritzo** [comment]
Maybe if `multiprocessing.cpu_count() == 1` we should execute synchronously rather than in parallel?

**Comment by neerajprad** [comment]
Currently, we would throw a warning in that case (I haven't seen such a warning on travis though), and set the number of processes to max cpu count. I just noticed though that we would still end up spawning a separate process rather than executing synchronously. I'll make that fix.

**Comment by fehiepsi** [comment]
@neerajprad Do you think that this causes `sv-dkl` fail? It's "maybe" to me. ^^

**Comment by neerajprad** [comment]
@fehiepsi - `sv-dkl` is just extremely slow. It doesn't freeze. You can see from the console output that it is progressing very slowly and hence times out because the entire stage takes longer than 50 mins. It also has a much higher failure rate (fails more than 50% of the times). This example on the other hand, doesn't fail often, but when it fails it just freezes and doesn't emit anything on to the console for 10 mins after which travis kills the job. My suspicion is that there is some deadlock that happens when we use multiprocessing. Other people have had issues with multiprocessing on travis, so I am pretty sure this example would work fine with `num_chains=1`.

**Comment by fritzo** [comment]
One pattern I've often used in the past is
```py
def parallel_map(fun, args, num_threads=None):
    args = list(args)
    if num_threads is None:
        num_threads = multiprocessing.cpu_count()
    if len(args) < 2 or num_threads == 1:
        return list(map(fun, args))
    return multiprocessing.Pool(num_threads).map(fun, args)

```

**Comment by fehiepsi** [comment]
@neerajprad I agree that it is extremely slow frequently in Travis recently. What I mean is that there might be some relevant to how much bandwidth Travis gives it in those cases. In our example tests, there are just two tests which consumes a lot of CPU cores: `sv-dkl` and `ss-vae`. It might not affect much other tests but for these two tests, they will be heavily affected. The thing is that `ss-vae` still passed, which makes me wondering. Because I can't replicate the problem locally in my machine, so all of these are just hypothesis. ^^

Btw, some times, Travis is pretty fast, e.g.: https://travis-ci.org/uber/pyro/builds/462873968?utm_source=github_status&utm_medium=notification

**Comment by neerajprad** [comment]
I expect that `baseball.py` example would be fixed with #1624. Regarding `sv-dkl`, I doubt if it is (solely) a travis issue because it was working without fail prior to pytorch-1.0 merge. There has to be some perf bug that we haven't isolated yet. :) On the other hand, `baseball.py` has been infrequently failing for a while even before the merge.

**Comment by neerajprad** [comment]
Closing this, please feel free to open if we notice this issue again.
